internal/repository/base.go

```diff
@@ -0,0 +1,485 @@
+		@applicationAddress)`
+
+	args := pgx.NamedArgs{
+		"firstBlock":           claim.FirstBlock,
+		"lastBlock":            claim.LastBlock,
+		"outputMerkleRootHash": claim.OutputMerkleRootHash,
+		"status":               claim.Status,
+		"applicationAddress":   claim.AppAddress,
+	}
+
+	_, err := pg.db.Exec(ctx, query, args)
+	if err != nil {
+		return fmt.Errorf("unable to insert row: %w\n", err)
+	}
+
+	return nil
+}
+
+func (pg *database) GetApplication(
+	ctx context.Context,
+	appAddress Address,
+) (*Application, error) {
+	var (
+		id                 uint64
+		templateHash       Hash
+		snapshotUri        string
+		lastProcessedBlock uint64
+		epochLength        uint64
+	)
+
+	query := `
+	SELECT
+		id,
+		template_hash,
+		snapshot_uri,
+		last_processed_block,
+		epoch_length
+	FROM
+		application
+	WHERE
+		application_address=$1`
+
+	err := pg.db.QueryRow(ctx, query, appAddress).Scan(&id, &templateHash, &snapshotUri, &lastProcessedBlock, &epochLength)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			slog.Info("GetApplication returned no rows", "service", "repository")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("GetApplication QueryRow failed: %v\n", err)
+	}
+
+	app := Application{
+		Id:                 id,
+		AppAddress:         appAddress,
+		TemplateHash:       templateHash,
+		SnapshotURI:        snapshotUri,
+		LastProcessedBlock: lastProcessedBlock,
+		EpochLength:        epochLength,
+	}
+
+	return &app, nil
+}
+
+func (pg *database) GetInput(
+	ctx context.Context,
+	index uint64,
+) (*Input, error) {
+	var (
+		id                 uint64
+		status             InputCompletionStatus
+		blob               []byte
+		blockNumber        uint64
+		machineHash        *Hash
+		outputsHash        *Hash
+		appAddress         Address
+		treatedMachineHash Hash
+		treatedOutputsHash Hash
+	)
+
+	query := `
+	SELECT
+		id,
+		blob,
+		status,
+		block_number,
+		machine_hash,
+		outputs_hash,
+		application_address
+	FROM
+		input
+	WHERE
+		index=$1`
+
+	err := pg.db.QueryRow(ctx, query, index).Scan(&id, &blob, &status, &blockNumber, &machineHash, &outputsHash, &appAddress)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			slog.Info("GetInput returned no rows", "service", "repository")
+			return nil, nil
+		}
+		return nil, fmt.Errorf("GetInput QueryRow failed: %v\n", err)
+	}
+
+	if machineHash == nil {
+		treatedMachineHash = common.HexToHash("0")
+	} else {
+		treatedMachineHash = *machineHash
+	}
+
+	if outputsHash == nil {
+		treatedOutputsHash = common.HexToHash("0")
+	} else {
+		treatedOutputsHash = *outputsHash
+	}
+
+	input := Input{
+		Id:               id,
+		Index:            index,
+		CompletionStatus: status,
+		Blob:             blob,
+		BlockNumber:      blockNumber,
+		MachineHash:      treatedMachineHash,
+		OutputsHash:      treatedOutputsHash,
+		AppAddress:       appAddress,
+	}
+
+	return &input, nil
+}
+
+func (pg *database) GetOutput(
+	ctx context.Context,
+	index uint64,
```

Comment:
In this case it is, output indexes are not repeatable and count from genesis

Reply:
I'm confused. Given two apps, one with two outputs, the other with three.
Won't we retrieve two rows if we select with index 1?

Reply:
We can align this in a call with other team members, as they may answer this more precisely, but from what I know it works like this:

The Cartesi Machine doesn't know (or care) about different apps. To it, it just get inputs and produces outputs, and when this output is produced, it's the next output, thus sequential indexes that are not repeatable. In your example, the outputs would be indexed 1-5, and if we want all the outputs from an app, we would need to use `application_address` in `input`.

Reply:
> The Cartesi Machine doesn't know (or care) about different apps
But the database does, right? If the goal of `GetOutput` is to return a single output, we must also specify the application it belongs to. Otherwise, we get the scenario that @renan061 described.

> the outputs would be indexed 1-5
Actually, they wouldn't. Each application has its output Merkle tree, so the output index is global in the context of a single application. It makes no sense to mix outputs from different apps together.